### PR TITLE
Add baseline Architecture to support auth + User sessions + Basic Auth

### PR DESCRIPTION
**Pull request by tianjing-li**
Beefy first PR to setup plug & play Auth:

- Adds new `auth` folder in services
- Adds `BaseAuthenticationStrategy` abstract class and `BasicAuthentication` (email+password) to start with. The base class should be inherited from all Auth strategies, next steps should include adding OAuth strategies.
- Adds `auth` routes, can retrieve session, logout, login, auth (to implement). The /login method should be generic and run for all auth strategies
- Adds `SessionMiddleware` to track current user session
- Adds extensive unit testing

- [x] **Add tests and docs**: Please include testing and documentation for your changes
- [x] **Lint and test**: Run `make lint` and `make run-tests`

### CODE REVIEW

File: src/backend/config/auth.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from backend.services.auth import BasicAuthentication
+
+ENABLED_AUTH_STRATEGIES = [
+    BasicAuthentication,
+]
```
Comment (tianjing-li) [inline src/backend/config/auth.py:5]: The config to set which auth strategies are active

File: src/backend/main.py [hunk 2]
```diff
@@ -15,32 +18,47 @@
...
     app.include_router(chat_router)
     app.include_router(user_router)
     app.include_router(conversation_router)
     app.include_router(tool_router)
     app.include_router(deployment_router)
     app.include_router(experimental_feature_router)
 
+    # Add auth
+    for auth in ENABLED_AUTH_STRATEGIES:
+        if auth.SHOULD_ATTACH_TO_APP:
```
Comment (tianjing-li) [inline src/backend/main.py:43]: This TODO is for setting up OAuth-based strategies, which will eventually require registering them to the FastAPI app during runtime

File: src/backend/main.py [hunk 3]
```diff
@@ -15,32 +18,47 @@
 
 load_dotenv()
 
+ORIGINS = ["*"]
```
Comment (rodrigue-h) [inline src/backend/main.py:21]: let's constrain this before merging - or allow someone to constrain via configuration?

Comment (tianjing-li) [reply src/backend/main.py:21]: I just capitalized the const, but we originally had `origins = ["*"]` set to streamline the deployments, without it a lot of the cloud deployments would require an extra step of modifying this in code, which we thought wasn't great UX

File: src/backend/services/auth/basic.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        Returns:
+            str: Hashed password
+        """
+        return bcrypt.hashpw(plain_text_password.encode("utf-8"), bcrypt.gensalt())
+
+    @staticmethod
+    def check_password(plain_text_password: str, hashed_password: str) -> bool:
+        """
+        Hashes a given plain-text password with a randomly generated salt.
```
Comment (scott-cohere) [inline src/backend/services/auth/basic.py:50]: i think this description might be an accidental copy paste from above

File: src/backend/services/auth/basic.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            plain_text_password (str): Password to check.
+            hashed_password (str): Password to check against.
+
+        Returns:
+            bool: Whether the plain-text password matches the given hashed password.
+        """
+        return bcrypt.checkpw(plain_text_password.encode("utf-8"), hashed_password)
+
+    @classmethod
+    def login(cls, session: Session, payload: dict[str, str]) -> bool:
```
Comment (scott-cohere) [inline src/backend/services/auth/basic.py:62]: return type should be dict | none

**Review by scott-cohere — APPROVED** [review]
just two tiny things. good stuff! 😃

File: src/backend/services/auth/basic.py (hunk 4 above)
Comment (tianjing-li) [reply src/backend/services/auth/basic.py:50]: ty

**Review by scott-cohere — APPROVED** [review]

**Review by scott-cohere — APPROVED** [review]

**Pull request merged**